internal/testutils/jimmtest/jimm.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2024 Canonical.
+
+package jimmtest
+
+import (
+	"time"
+
+	"github.com/google/uuid"
+
+	"github.com/canonical/jimm/v3/internal/db"
+	"github.com/canonical/jimm/v3/internal/jimm"
+	"github.com/canonical/jimm/v3/internal/jimmjwx"
+	"github.com/canonical/jimm/v3/internal/pubsub"
+)
+
+var now = (time.Time{}).UTC().Round(time.Millisecond)
+
+type Option func(j *jimm.JIMM)
+
+var (
+	UnsetCredentialStore Option = func(j *jimm.JIMM) {
+		j.CredentialStore = nil
+	}
+)
+
+func NewJIMM(t Tester, additionalParameters *jimm.Parameters, options ...Option) *jimm.JIMM {
+
+	auth := NewMockOAuthAuthenticator(t, nil)
+
+	p := jimm.Parameters{
+		UUID:               uuid.NewString(),
+		Dialer:             &Dialer{},
+		Pubsub:             &pubsub.Hub{},
+		JWKService:         &jimmjwx.JWKSService{},
+		JWTService:         &jimmjwx.JWTService{},
+		OAuthAuthenticator: &auth,
+	}
+
+	if additionalParameters != nil {
+		if additionalParameters.UUID != "" {
+			p.UUID = additionalParameters.UUID
+		}
+		if additionalParameters.Dialer != nil {
+			p.Dialer = additionalParameters.Dialer
+		}
+		if additionalParameters.Database != nil {
+			p.Database = additionalParameters.Database
+		}
+		if additionalParameters.CredentialStore != nil {
+			p.CredentialStore = additionalParameters.CredentialStore
+		}
+		if additionalParameters.Pubsub != nil {
+			p.Pubsub = additionalParameters.Pubsub
+		}
+		if len(additionalParameters.ReservedCloudNames) > 0 {
+			p.ReservedCloudNames = append(p.ReservedCloudNames, additionalParameters.ReservedCloudNames...)
+		}
+		if additionalParameters.OpenFGAClient != nil {
+			p.OpenFGAClient = additionalParameters.OpenFGAClient
+		}
+		if additionalParameters.JWKService != nil {
+			p.JWKService = additionalParameters.JWKService
+		}
+		if additionalParameters.JWTService != nil {
+			p.JWTService = additionalParameters.JWTService
+		}
+		if additionalParameters.OAuthAuthenticator != nil {
+			p.OAuthAuthenticator = additionalParameters.OAuthAuthenticator
+		}
+	}
+
+	if p.Database == nil {
+		p.Database = &db.Database{
+			DB: PostgresDB(t, func() time.Time { return now }),
```

Comment:
I'm currently not in favour of this approach where every test that calls `jimmtest.NewJIMM` gets a whole database and OpenFGA store. Some tests need neither, some tests need one. That was nicely captured in the tests and when we move to managers we can use suites so that there is less duplication of that boilerplate test setup.

I think the default should be all mocks or all nil fields unless the test requests them.